### PR TITLE
Go: read tree

### PR DESCRIPTION
**Pull request by nikandfor**

### CODE REVIEW

**Comment by nikandfor** [comment]
I've implemented two ls-tree functions. One uses bufio.Reader, and one uses bytes.IndexByte. I would personally prefer the latter as bufio.Reader doesn't really simplify the code but instead adds one more unnecessary layer of abstraction. @rohitpaulk, which one looks better to you?

**Comment by nikandfor** [comment]
I've also added `object` package and moved its operations there. I think this is the perfect time to separate object concept and operations with it.

**Comment by rohitpaulk** [comment]
@nikandfor out of the two options I'd go with `bufio.Reader` since the operations seem more high-level and readable to me. Agree that in terms of overall code complexity it doesn't make a huge dent compared to the IndexByte implementation.

Adding inline comments for the rest...

**Review by rohitpaulk — APPROVED** [review]

File: solutions/go/read_tree/code/object/load.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		e := file.Close()
+		if err == nil && e != nil {
+			err = fmt.Errorf("close file: %w", e)
+		}
+	}()
+
+	return Load(file)
+}
+
+func Load(r io.Reader) (typ string, content []byte, err error) {
```
[the remark this answers is not in the document]
Comment (nikandfor) [reply solutions/go/read_tree/code/object/load.go:40]: Yes, it's for potential lower level operations. Load* methods are all siblings. They load objects by different features: by hash or from raw reader. For example we may open file manually and read it from command like `git inspect-file .git/objects/xx/yyy...yyy`

[the remark this answers is not in the document]
Comment (rohitpaulk) [reply solutions/go/read_tree/code/object/load.go:40]: Got it - yeah sounds fine to me 👍

**Pull request merged**